[gameplay]/ladders/client.lua

```diff
@@ -0,0 +1,968 @@
+-- ( xxxx |client) by: Markland aka DarkChemical
+
+
+-- ** Data ** --
+local climbers = {}
+local climbs = {}
+
+local anim = anims
+
+
+
+
+addEvent("onClientPedLadderClimbingStart")
+addEvent("onClientPedLadderClimbingStop")
+addEvent("onClientPedLadderClimbingStep")
+
+addEvent("onClientRecieveLadderClimbingReady", true)
+addEvent("onClientRecievePedLadderClimbingState", true)
+addEvent("onClientRecieveLadderState", true)
+
+
+function getLineUnitScale3D(sx, sy, sz, ex, ey, ez, m)
+	local m = m or 1
+	local sx, sy, sz, ex, ey, ez = tonumber(sx) or 0, tonumber(sy) or 0, tonumber(sz) or 0, tonumber(ex) or 0, tonumber(ey) or 0, tonumber(ez) or 0
+	local ax, ay, az = ex-sx, ey-sy, ez-sz
+	local d = getDistanceBetweenPoints3D(0, 0, 0, ax, ay, az)
+	if d == 0 or not d then return 0, 0, 0 end
+	local dx, dy, dz = ax/d, ay/d, az/d
+	return dx*m, dy*m, dz*m
+end
+
+function getPointPositionOnLine3D(px, py, pz, sx, sy, sz, tx, ty, tz)
+	local st = getDistanceBetweenPoints3D(sx, sy, sz, tx, ty, tz)
+	local sp = getDistanceBetweenPoints3D(sx, sy, sz, px, py, pz)
+	local tp = getDistanceBetweenPoints3D(tx, ty, tz, px, py, pz)
+	local atp = math.acos((sp^2+st^2-tp^2)/(2*sp*st))
+	local sd = sp*math.cos(atp)
+	if sd>st or sd<0 then
+		if tp<sp then
+			return tx, ty, tz
+		else
+			return sx, sy, sz
+		end
+	else
+		local x, y, z = getLineUnitScale3D(sx, sy, sz, tx, ty, tz, sd)
+		return sx+x, sy+y, sz+z
+	end
+end
+
+
+function getPointDistanceFromLine3D(px, py, pz, sx, sy, sz, tx, ty, tz)
+	local st = getDistanceBetweenPoints3D(sx, sy, sz, tx, ty, tz)
+	local sp = getDistanceBetweenPoints3D(sx, sy, sz, px, py, pz)
+	local tp = getDistanceBetweenPoints3D(tx, ty, tz, px, py, pz)
+	local atp = math.acos((sp^2+st^2-tp^2)/(2*sp*st))
+	local sd = sp*math.cos(atp)
+	if sd>st or sd<0 then
+		return tp<sp and tp or sp
+	else
+		return (sp^2-sd^2)^.5
+	end
+end
+
+
+function getPointDistanceFromLadder(px, py, pz, sx, sy, sz, tx, ty, tz)
+	local st = getDistanceBetweenPoints3D(sx, sy, sz, tx, ty, tz)
+	local sp = getDistanceBetweenPoints3D(sx, sy, sz, px, py, pz)
+	local tp = getDistanceBetweenPoints3D(tx, ty, tz, px, py, pz)
+	local atp = math.acos((sp^2+st^2-tp^2)/(2*sp*st))
+	local sd = sp*math.cos(atp)
+	if sd>st or sd<0 then
+		return tp<sp and tp or sp, tp<sp and 1 or 0
+	else
+		return (sp^2-sd^2)^.5, sd/st
+	end
+end
+
+
+local minus
+do
+	local abs = math.abs
+	-- calculate and return the distance unit2 
+	-- needs to move to reach unit1
+	function minus(unit1, unit2)
+		local phi = abs(unit2-unit1) % 360
+		local sign = 1
+		-- used to calculate sign
+		if not ((unit1-unit2 >= 0 and unit1-unit2 <= 180)
+		or (unit1-unit2 <= -180 and unit1-unit2 >= -360)) then
+			sign = -1
+		end
+		if phi > 180 then
+			result = 360-phi
+		else
+			result = phi
+		end
+
+		return result*sign
+	end
+end
```

Comment:
This can be implemented much more simply, how about this solution?
```suggestion
function minus(unit1, unit2)
    local diff = unit2 - unit1
    local diffNormalized = diff % 360
    
    return diffNormalized > 180 and diffNormalized - 360 or diffNormalized
end
```

Reply:
Looks good, don't see why not as long as it works.